standards/api_security_standards.md

```diff
@@ -0,0 +1,158 @@
+
+1.3 Secure Object Identifiers: Use random and unpredictable values, such as GUIDs, for object identifiers to prevent attackers from guessing and manipulating object IDs.
+
+1.4 Testing and Verification: Regularly test the authorization mechanisms to identify and rectify any flaws. Ensure that changes that could introduce vulnerabilities are not deployed without thorough testing.
+
+**2. Broken Authentication**
+
+Authentication mechanisms are often implemented incorrectly, allowing attackers to compromise authentication tokens or to exploit implementation flaws to assume other user's identities temporarily or permanently. Compromising system's ability to identify the client/user, compromises API security overall.
+
+2.1 Robust Authentication Flows: Ensure all authentication flows are thoroughly designed and tested, including mobile, web, and any deep links that implement one-click authentication.
+
+2.2 Endpoint Security: Treat all authentication endpoints, including credential recovery and password reset, with high-security measures such as rate limiting, captcha, and account lockout mechanisms.
+
+2.3 Strong Authentication Practices: Utilize industry-standard methods for authentication, token generation, and password storage. Avoid reinventing these mechanisms.
+
+2.4 Multi-Factor Authentication (MFA): Implement multi-factor authentication wherever possible to enhance security.
+
+2.5 Anti-Brute Force Measures: Deploy strict anti brute force measures to protect against credential stuffing and dictionary attacks. These should be more stringent than regular rate limiting on other API endpoints.
+
+**3. Excessive Data Exposure**
+
+Looking forward to generic implementations, developers tend to expose all object properties without considering their individual sensitivity, relying on clients to perform the data filtering before displaying it to the user.
+
+3.1 Client-Side Filtering Prohibition: Ensure that no sensitive data is ever only filtered on the client side. All data filtering and reduction to necessary elements must occur server-side before transmission.
+
+3.2 Response Review and Restriction: Regularly review API responses to ensure they contain only data that is appropriate and necessary for the client. Avoid returning excessive or sensitive information that exceeds the requirements of the API consumer.
+
+3.3 Property-specific Responses: Refrain from using generic serialization methods like toJSON() for objects that may contain sensitive data. Instead, explicitly define and return only the necessary properties in API responses.
+
+3.4 Data Classification and Schema Validation: Classify all data handled by the application, particularly identifying any personally identifiable information (PII). Implement schema-based validation for all API responses to ensure compliance with data exposure policies.
```

Comment:
Let's add a sample snippet? We can add insecure and secure example for 3.1 to 3.4 @arcubalino

Reply:
As discussed, the security standards are generic and can be applied to any programming language.

Resolution: We won't be adding snippets